src/python/pants/engine/legacy/structs.py

```diff
@@ -325,17 +337,12 @@ def from_sources_field(
 
   @staticmethod
   def _filespec_for_exclude(raw_exclude: List[str], spec_path: str) -> wrapped_globs._GlobsDict:
-    if isinstance(raw_exclude, str):
```

Comment:
Is there a reason we moved the validation out of this method?

Reply:
Yes. We rewrote the call site so that `raw_exclude=[*self._exclude_patterns, *(exclude or [])]`. This means that `raw_exclude` will now always be a list, so the exception would never raise.

It's safe to simply move it into the `__init__()` for `BaseGlobs` as this is a private method and no other function calls this method.

(Sorry I didn't proactively explain this 😬)

Reply:
Ah! Being able to trust the types here is really really nice!

Reply:
> (Sorry I didn't proactively explain this 😬)

No worries! I think I would leave a tiny note about this on the PR in the future, but it's not a mistake to not do it, someone like me will just ask you about it if it needs asking!